File: Code/pit_hd110067_sim/pit_vacuum_yee_test.jl
"""
PIT Vacuum Wave Test — 2D Yee-grid FDTD (Julia)

Goals
- Correct Yee staggering (Ex: Nx×(Ny-1), Ey: (Nx-1)×Ny, Bz: (Nx-1)×(Ny-1)).
- Stable run for ≥800 steps (no NaN/Inf) with sponge + mild damping.
- Wave speed near theory (c_theory = sqrt(lambda_vac/gamma_vac)).
- Outputs:
    pit_yee_wavefront_tracking_stable.png
    pit_yee_propagation_heatmap_stable.png

Note: plotting uses PyPlot; ensure it is available in your Julia environment.
"""

using LinearAlgebra
using Statistics
using DelimitedFiles
HAVE_PYPLOT = false
try
    using PyPlot
    global HAVE_PYPLOT = true
catch
    @warn "PyPlot not available; skipping PNG generation."
end

# -----------------------------
# Parameters
# -----------------------------
struct VacuumParams
    lambda_vac::Float64
    gamma_vac::Float64
    dx::Float64
    dt::Float64
    Nx::Int
    Ny::Int
    steps::Int
    sponge_width::Int
    sponge_strength::Float64
    global_damp::Float64
    source_amp::Float64
    source_w_space::Float64
    source_w_time::Float64
    source_t0::Float64
end

function VacuumParams(; lambda_vac=1.0, gamma_vac=1.0, dx=1.0, Nx=201, Ny=201, steps=900,
    sponge_width=25, sponge_strength=0.03, global_damp=0.999,
    source_amp=0.05, source_w_space=4.0, source_w_time=15.0, source_t0=45.0,
    dt=nothing)
    c_th = sqrt(lambda_vac / gamma_vac)
    dt === nothing && (dt = 0.30 * dx / (c_th * sqrt(2))) # conservative CFL
    VacuumParams(lambda_vac, gamma_vac, dx, dt, Nx, Ny, steps, sponge_width,
        sponge_strength, global_damp, source_amp, source_w_space, source_w_time, source_t0)
end

epsilon_eff(p::VacuumParams) = 1 / p.lambda_vac
mu_eff(p::VacuumParams) = p.gamma_vac
c_theory(p::VacuumParams) = sqrt(p.lambda_vac / p.gamma_vac)

# -----------------------------
# Field allocation
# -----------------------------
struct YeeState
    Ex::Matrix{Float64}  # (Nx, Ny-1)
    Ey::Matrix{Float64}  # (Nx-1, Ny)
    Bz::Matrix{Float64}  # (Nx-1, Ny-1)
end

function init_fields(p::VacuumParams)
    Ex = zeros(p.Nx, p.Ny - 1)
    Ey = zeros(p.Nx - 1, p.Ny)
    Bz = zeros(p.Nx - 1, p.Ny - 1)
    YeeState(Ex, Ey, Bz)
end

# -----------------------------
# Sponge / damping masks
# -----------------------------
function make_sponge(p::VacuumParams)
    Xc = repeat(collect(0:p.Nx-2), 1, p.Ny - 1)
    Yc = repeat(collect(0:p.Ny-2)', p.Nx - 1, 1)
    dist = min.(Xc, Yc, (p.Nx - 2) .- Xc, (p.Ny - 2) .- Yc)
    sponge_B = ones(p.Nx - 1, p.Ny - 1)
    edge = dist .< p.sponge_width
    sponge_B[edge] .= exp.(-p.sponge_strength .* (p.sponge_width .- dist[edge]))

    sponge_Ex = ones(p.Nx, p.Ny - 1)
    sponge_Ex[1:end-1, :] .= sponge_B

    sponge_Ey = ones(p.Nx - 1, p.Ny)
    sponge_Ey[:, 1:end-1] .= sponge_B
    return sponge_Ex, sponge_Ey, sponge_B
end

# -----------------------------
# Source
# -----------------------------
function inject_source!(Bz, n, p::VacuumParams, cx, cy)
    g_t = p.source_amp * exp(-((n - p.source_t0) / p.source_w_time)^2)
    x = collect(0:size(Bz, 1)-1) .- cx
    y = collect(0:size(Bz, 2)-1) .- cy
    X = repeat(x, 1, length(y))
    Y = repeat(y', length(x), 1)
    g_xy = exp.(-((X .^ 2 .+ Y .^ 2) ./ (2 * p.source_w_space^2)))
    @. Bz += g_t * g_xy
end

# -----------------------------
# Updates (Yee staggered)
# -----------------------------
function update_E!(state::YeeState, p::VacuumParams)
    dBz_dy = (state.Bz[:, 2:end] .- state.Bz[:, 1:end-1]) ./ p.dx        # (Nx-1, Ny-2)
    state.Ex[1:end-1, 2:end] .+= (p.dt / epsilon_eff(p)) .* dBz_dy

    dBz_dx = (state.Bz[2:end, :] .- state.Bz[1:end-1, :]) ./ p.dx        # (Nx-2, Ny-1)
    state.Ey[2:end, 1:end-1] .-= (p.dt / epsilon_eff(p)) .* dBz_dx
end

function update_B!(state::YeeState, p::VacuumParams)
    dEy_dx = (state.Ey[2:end, 1:end-1] .- state.Ey[1:end-1, 1:end-1]) ./ p.dx  # (Nx-2, Ny-1)
    dEx_dy = (state.Ex[1:end-1, 2:end] .- state.Ex[1:end-1, 1:end-1]) ./ p.dx  # (Nx-1, Ny-2)
    curlE = dEy_dx[:, 1:end-1] .- dEx_dy[1:end-1, :]                            # (Nx-2, Ny-2)
    state.Bz[2:end, 2:end] .-= (p.dt / mu_eff(p)) .* curlE
end

function apply_damping!(state::YeeState, sponge_Ex, sponge_Ey, sponge_B, p::VacuumParams)
    @. state.Ex = state.Ex * sponge_Ex * p.global_damp
    @. state.Ey = state.Ey * sponge_Ey * p.global_damp
    @. state.Bz = state.Bz * sponge_B * p.global_damp
end

# -----------------------------
# Simulation
# -----------------------------
function run_simulation(p::VacuumParams)
    state = init_fields(p)
    sponge_Ex, sponge_Ey, sponge_B = make_sponge(p)
    cx, cy = (p.Nx - 2) ÷ 2, (p.Ny - 2) ÷ 2  # center on Bz grid

    # radial diagnostics on Bz grid
    x = collect(0:p.Nx-2) .- cx
    y = collect(0:p.Ny-2) .- cy
    X = repeat(x, 1, length(y))
    Y = repeat(y', length(x), 1)
    R = sqrt.(X .^ 2 .+ Y .^ 2)
    max_r = floor(Int, maximum(R))
    shells = collect(1:max_r)
    shell_masks = [(R .>= r - 0.5) .& (R .< r + 0.5) for r in shells]

    heatmap = zeros(p.steps, length(shells))
    wavefront = zeros(p.steps)
    radial_arrival = Dict{Int,Int}()
    threshold = 1e-3

    for n in 1:p.steps
        if n <= 140
            inject_source!(state.Bz, n, p, cx, cy)
        end

        update_E!(state, p)
        update_B!(state, p)
        apply_damping!(state, sponge_Ex, sponge_Ey, sponge_B, p)

        shell_vals = similar(shells, Float64)
        @inbounds for (i, mask) in enumerate(shell_masks)
            val = maximum(abs.(state.Bz[mask]))
            heatmap[n, i] = val
            shell_vals[i] = val
            if !haskey(radial_arrival, shells[i]) && val > threshold
                radial_arrival[shells[i]] = n
            end
        end

        if sum(shell_vals) > 0
            wavefront[n] = shells[argmax(shell_vals)]
        end

        if any(isnan, state.Bz) || any(isinf, state.Bz)
            error("Numerical instability at step $n")
        end
    end

    return state, heatmap, wavefront, radial_arrival, shells
end

# -----------------------------
# Wave speed fit and plots
# -----------------------------
function fit_wave_speed_from_series(r_fit::Vector{Float64}, t_fit::Vector{Float64}, c_th::Float64)
    slope_rt, intercept_r = fit_line(t_fit, r_fit) # r = v*t + b
    c_meas = slope_rt
    rel_err = abs(c_meas - c_th) / c_th
    slope_t = 1 / slope_rt
    intercept_t = -intercept_r / slope_rt
    return c_meas, slope_t, intercept_t, rel_err
end

fit_line(x, y) = begin
    X = [ones(length(x)) x]
    coef = X \ y
    intercept, slope = coef[1], coef[2]
    return slope, intercept
end

function plot_wavefront_tracking(r_vals, t_vals, slope_t, intercept_t, c_measured)
    HAVE_PYPLOT || return
    figure(figsize=(7,5))
    scatter(r_vals, t_vals, s=12, label="Wavefront radii")
    plot(r_vals, slope_t .* r_vals .+ intercept_t, lw=2, label="Fit: c=$(round(c_measured, digits=4))")
    title("PIT Vacuum Wavefront Tracking (Yee Grid, stable)")
    xlabel("Radius (node index)")
    ylabel("Time (steps)")
    grid(true)
    legend()
    tight_layout()
    savefig("pit_yee_wavefront_tracking_stable.png", dpi=160)
    close()
end

function plot_heatmap(heatmap, shells)
    HAVE_PYPLOT || return
    figure(figsize=(8,4))
    imshow(permutedims(heatmap), origin="lower", aspect="auto",
           extent=[1, size(heatmap, 1), shells[1], shells[end]])
    colorbar(label="|Bz| amplitude")
    title("PIT Vacuum Wave Propagation Heatmap (Yee Grid, stable)")
    xlabel("Time step")
    ylabel("Radius (nodes)")
    tight_layout()
    savefig("pit_yee_propagation_heatmap_stable.png", dpi=160)
    close()
end

# -----------------------------
# Main
# -----------------------------
function main()
    p = VacuumParams()
    println("\n--- PIT Yee Vacuum Test (Julia) ---")
    println("lambda_vac=$(p.lambda_vac), gamma_vac=$(p.gamma_vac)")
    println("epsilon_eff=1/lambda=$(1/p.lambda_vac), mu_eff=gamma=$(p.gamma_vac)")
    println("c_theory=$(c_theory(p)) nodes/step")
    println("dt=$(p.dt) (CFL guard 0.30/sqrt(2))")
    println("steps=$(p.steps)\n")

    state, heatmap, wavefront, radial_arrival, shells = run_simulation(p)

    r_sorted = sort(collect(keys(radial_arrival)))
    t_sorted = [radial_arrival[r] for r in r_sorted]
    if length(r_sorted) < 10
        error("Wavefront detection failed; insufficient arrival data.")
    end
    slope_t, intercept_t = fit_line(r_sorted, t_sorted)   # t = slope*r + b
    c_measured = 1 / slope_t
    rel_err = abs(c_measured - c_theory(p)) / c_theory(p)
    r_fit = Float64.(r_sorted)
    t_fit = Float64.(t_sorted)

    println("--- Results ---")
    println("c_measured = $(round(c_measured, digits=6)) nodes/step")
    println("relative error = $(round(rel_err*100, digits=2))%")

    for (name, arr) in [("Ex", state.Ex), ("Ey", state.Ey), ("Bz", state.Bz)]
        any(isnan, arr) && error("$name has NaN")
        any(isinf, arr) && error("$name has Inf")
    end

    plot_wavefront_tracking(r_fit, t_fit, slope_t, intercept_t, c_measured)
    plot_heatmap(heatmap, shells)

    # CSV outputs for offline plotting
    writedlm("pit_yee_wavefront_series.csv", [t_fit r_fit], ',')
    writedlm("pit_yee_heatmap.csv", heatmap, ',')
    r_sorted = sort(collect(keys(radial_arrival)))
    t_sorted = [radial_arrival[r] for r in r_sorted]
    writedlm("pit_yee_radial_arrivals.csv", [r_sorted t_sorted], ',')

    println("Wrote:")
    println("  pit_yee_wavefront_series.csv")
    println("  pit_yee_heatmap.csv")
    println("  pit_yee_radial_arrivals.csv")
    if HAVE_PYPLOT
        println("  pit_yee_wavefront_tracking_stable.png")
        println("  pit_yee_propagation_heatmap_stable.png")
    else
        println("Skipped PNGs (PyPlot not available).")
    end
end

main()

File: Code/pit_hd110067_sim/pit_vacuum_yee_test.py
"""
PIT Vacuum Wave Test — 2D Yee Grid (TEz)

Goals:
- Correct Yee staggering with matching derivative shapes (no broadcasting).
- Stable for >=800 steps (no NaN/inf).
- Wave speed close to theory.
- Outputs:
    pit_yee_wavefront_tracking_stable.png
    pit_yee_propagation_heatmap_stable.png
"""

from dataclasses import dataclass
import numpy as np
import matplotlib.pyplot as plt


@dataclass
class VacuumParams:
    lambda_vac: float = 1.0
    gamma_vac: float = 1.0
    dx: float = 1.0
    dt: float | None = None
    Nx: int = 201
    Ny: int = 201
    steps: int = 900
    sponge_width: float = 30.0
    sponge_strength: float = 0.03
    global_damp: float = 0.9995
    diffusion: float = 0.002
    source_amp: float = 0.05
    source_w_space: float = 4.0
    source_w_time: float = 15.0
    source_t0: float = 45.0

    def __post_init__(self):
        self.epsilon_eff = 1.0 / self.lambda_vac
        self.mu_eff = self.gamma_vac
        self.c_theory = np.sqrt(self.lambda_vac / self.gamma_vac)
        if self.dt is None:
            # Conservative CFL to improve stability.
            self.dt = 0.55 * self.dx / (self.c_theory * np.sqrt(2))


@dataclass
class YeeState:
    Ex: np.ndarray  # (Nx,   Ny-1)
    Ey: np.ndarray  # (Nx-1, Ny)
    Bz: np.ndarray  # (Nx-1, Ny-1)


def initialize_fields(params: VacuumParams) -> YeeState:
    Ex = np.zeros((params.Nx, params.Ny - 1), dtype=float)
    Ey = np.zeros((params.Nx - 1, params.Ny), dtype=float)
    Bz = np.zeros((params.Nx - 1, params.Ny - 1), dtype=float)
    return YeeState(Ex, Ey, Bz)


def make_sponge(params: VacuumParams):
    Xc = np.arange(params.Nx - 1)
    Yc = np.arange(params.Ny - 1)
    XXc, YYc = np.meshgrid(Xc, Yc, indexing="ij")
    dist = np.minimum.reduce([XXc, YYc, (params.Nx - 2 - XXc), (params.Ny - 2 - YYc)]).astype(float)

    sponge_B = np.ones((params.Nx - 1, params.Ny - 1), dtype=float)
    edge = dist < params.sponge_width
    sponge_B[edge] = np.exp(-params.sponge_strength * (params.sponge_width - dist[edge]))

    sponge_Ex = np.ones((params.Nx, params.Ny - 1), dtype=float)
    sponge_Ex[:-1, :] = sponge_B

    sponge_Ey = np.ones((params.Nx - 1, params.Ny), dtype=float)
    sponge_Ey[:, :-1] = sponge_B
    return sponge_Ex, sponge_Ey, sponge_B


def inject_source(Bz: np.ndarray, n: int, params: VacuumParams, cx: int, cy: int):
    g_t = params.source_amp * np.exp(-((n - params.source_t0) / params.source_w_time) ** 2)
    x = np.arange(Bz.shape[0]) - cx
    y = np.arange(Bz.shape[1]) - cy
    X, Y = np.meshgrid(x, y, indexing="ij")
    g_xy = np.exp(-((X**2 + Y**2) / (2 * params.source_w_space**2)))
    Bz += g_t * g_xy


def update_E_fields(state: YeeState, params: VacuumParams):
    # Ex update from dBz/dy -> shape (Nx-1, Ny-2)
    dBz_dy = (state.Bz[:, 1:] - state.Bz[:, :-1]) / params.dx
    state.Ex[:-1, 1:] += (params.dt / params.epsilon_eff) * dBz_dy

    # Ey update from dBz/dx -> shape (Nx-2, Ny-1)
    dBz_dx = (state.Bz[1:, :] - state.Bz[:-1, :]) / params.dx
    state.Ey[1:-1, :-1] -= (params.dt / params.epsilon_eff) * dBz_dx


def update_B_field(state: YeeState, params: VacuumParams):
    # curlE on Bz grid: dEy/dx - dEx/dy
    dEy_dx = (state.Ey[1:, :-1] - state.Ey[:-1, :-1]) / params.dx  # (Nx-2, Ny-1)
    dEx_dy = (state.Ex[:-1, 1:] - state.Ex[:-1, :-1]) / params.dx  # (Nx-1, Ny-2)
    curlE = dEy_dx[:, :-1] - dEx_dy[:-1, :]                         # (Nx-2, Ny-2)
    state.Bz[1:-1, 1:-1] -= (params.dt / params.mu_eff) * curlE


def apply_damping(state: YeeState, sponge_Ex, sponge_Ey, sponge_B, params: VacuumParams):
    state.Ex *= sponge_Ex * params.global_damp
    state.Ey *= sponge_Ey * params.global_damp
    state.Bz *= sponge_B * params.global_damp

    if params.diffusion > 0:
        B = state.Bz
        Bp = np.pad(B, ((1, 1), (1, 1)), mode="edge")
        lap = (Bp[2:, 1:-1] + Bp[:-2, 1:-1] + Bp[1:-1, 2:] + Bp[1:-1, :-2] - 4 * B)
        state.Bz = B + params.diffusion * lap


def run_simulation(params: VacuumParams):
    state = initialize_fields(params)
    sponge_Ex, sponge_Ey, sponge_B = make_sponge(params)
    cx, cy = (params.Nx - 2) // 2, (params.Ny - 2) // 2  # Bz-centered

    x = np.arange(params.Nx - 1) - cx
    y = np.arange(params.Ny - 1) - cy
    X, Y = np.meshgrid(x, y, indexing="ij")
    R = np.sqrt(X**2 + Y**2)
    max_r = int(R.max())
    shells = np.arange(1, max_r + 1)
    shell_masks = [(R >= r - 0.5) & (R < r + 0.5) for r in shells]

    heatmap = np.zeros((params.steps, len(shells)), dtype=float)
    radial_arrival: dict[int, int] = {}
    wavefront_series = np.zeros(params.steps, dtype=float)
    threshold = 2e-3

    for n in range(params.steps):
        if n < 140:
            inject_source(state.Bz, n, params, cx, cy)

        update_E_fields(state, params)
        update_B_field(state, params)
        apply_damping(state, sponge_Ex, sponge_Ey, sponge_B, params)

        shell_vals = []
        for i, (r, mask) in enumerate(zip(shells, shell_masks)):
            val = float(np.max(np.abs(state.Bz[mask])))
            heatmap[n, i] = val
            if r not in radial_arrival and val > threshold:
                radial_arrival[r] = n
            shell_vals.append(val)

        if np.sum(shell_vals) > 0:
            wavefront_series[n] = shells[int(np.argmax(shell_vals))]

        if np.isnan(state.Bz).any() or np.isinf(state.Bz).any():
            raise RuntimeError(f"Numerical instability at step {n}")

    return state, heatmap, radial_arrival, wavefront_series, shells


def fit_wave_speed_from_series(r_fit: np.ndarray, t_fit: np.ndarray, c_theory: float):
    slope_rt, intercept_r = np.polyfit(t_fit, r_fit, 1)  # r = v*t + b
    c_measured = slope_rt
    rel_err = abs(c_measured - c_theory) / c_theory
    slope_t = 1.0 / slope_rt if slope_rt != 0 else np.inf
    intercept_t = -intercept_r / slope_rt if slope_rt != 0 else 0.0
    return c_measured, slope_t, intercept_t, rel_err


def plot_wavefront_tracking(r_vals, t_vals, slope_t, intercept_t, c_measured):
    plt.figure(figsize=(7, 5))
    plt.scatter(r_vals, t_vals, s=12, label="Wavefront radii")
    plt.plot(r_vals, slope_t * r_vals + intercept_t, lw=2, label=f"Fit: c_sim={c_measured:.3f}")
    plt.title("PIT Vacuum Wavefront Tracking (Yee Grid, stable)")
    plt.xlabel("Radius (node index)")
    plt.ylabel("Time (steps)")
    plt.grid(True)
    plt.legend()
    plt.tight_layout()
    plt.savefig("pit_yee_wavefront_tracking_stable.png", dpi=160)
    plt.close()


def plot_heatmap(heatmap, shells):
    plt.figure(figsize=(8, 4))
    plt.imshow(
        heatmap.T,
        origin="lower",
        aspect="auto",
        extent=[0, heatmap.shape[0], shells[0], shells[-1]],
    )
    plt.colorbar(label="|Bz| amplitude")
    plt.title("PIT Vacuum Wave Propagation Heatmap (Yee Grid, stable)")
    plt.xlabel("Time step")
    plt.ylabel("Radius (nodes)")
    plt.tight_layout()
    plt.savefig("pit_yee_propagation_heatmap_stable.png", dpi=160)
    plt.close()


def main():
    params = VacuumParams()
    print("\n--- PIT Yee Vacuum Test (stable) ---")
    print(f"lambda_vac={params.lambda_vac}, gamma_vac={params.gamma_vac}")
    print(f"epsilon_eff=1/lambda={params.epsilon_eff:.6g}, mu_eff=gamma={params.mu_eff:.6g}")
    print(f"c_theory=sqrt(lambda/gamma)={params.c_theory:.6f} nodes/step")
    print(f"dt = {params.dt:.6f} (CFL guard 0.55/sqrt(2))")
    print(f"steps = {params.steps}\n")

    state, heatmap, radial_arrival, wavefront_series, shells = run_simulation(params)

    valid = np.where(wavefront_series > 0)[0]
    if len(valid) < 10:
        raise RuntimeError("Wavefront detection failed; insufficient data.")
    t_fit = valid
    r_fit = wavefront_series[valid]
    c_measured, slope_t, intercept_t, rel_err = fit_wave_speed_from_series(r_fit, t_fit, params.c_theory)

    print("--- Results ---")
    print(f"c_measured = {c_measured:.6f} nodes/step")
    print(f"slope = {slope_t:.6f} steps/node")
    print(f"relative error = {rel_err*100:.2f}%\n")

    for name, arr in [("Ex", state.Ex), ("Ey", state.Ey), ("Bz", state.Bz)]:
        assert not np.isnan(arr).any(), f"{name} has NaN"
        assert not np.isinf(arr).any(), f"{name} has inf"

    plot_wavefront_tracking(r_fit, t_fit, slope_t, intercept_t, c_measured)
    plot_heatmap(heatmap, shells)

    print("Wrote:")
    print("  pit_yee_wavefront_tracking_stable.png")
    print("  pit_yee_propagation_heatmap_stable.png")


if __name__ == "__main__":
    main()
